queasars/job_shop_scheduling/scip_solver.py

```diff
@@ -0,0 +1,136 @@
+# Quantum Evolving Ansatz Variational Solver (QUEASARS)
+# Copyright 2024 DLR - Deutsches Zentrum für Luft- und Raumfahrt e.V.
+
+from typing import Optional
+from itertools import combinations
+
+from pyscipopt import Model
+from pyscipopt.scip import Variable, Solution
+
+from queasars.job_shop_scheduling.problem_instances import (
+    Machine,
+    Job,
+    Operation,
+    JobShopSchedulingProblemInstance,
+    JobShopSchedulingResult,
+    ScheduledOperation,
+)
+
+
+class JSSPSCIPModelEncoder:
+    """
+    Encoding class used to encode a JobShopSchedulingProblemInstance as a pyscipopt SCIP solver model.
+    Needs queasars to be installed with the pyscipopt extra: pip install queasars[pyscipopt].
+    It also needs SCIP to be installed separately in the version 8.*.
+
+    :param jssp_instance: job shop scheduling problem instance to encode as a scip model
+    :type jssp_instance: JobShopSchedulingProblemInstance
+    """
+
+    def __init__(self, jssp_instance: JobShopSchedulingProblemInstance):
+        self._jssp_instance: JobShopSchedulingProblemInstance = jssp_instance
+        self._model_prepared: bool = False
+        self._model: Optional[Model] = None
+        self._optimization_var: Optional[Variable] = None
+        self._operation_start_variables: dict[Operation, Variable] = {}
+        self._machine_operations: dict[Machine, list[Operation]] = {}
```

Comment:
Yes all this could be setup in the constructor already. As a result, i moved some of the initialization into the constructor. The other parts i kept in the methods _prepare_variables() and _prepare_constraints to not overload the constructor. This also has the benefit of only setting up the variables and constraints once the user requests the scip model.